### PR TITLE
`bp.jl_bitround`

### LINKED ISSUE #25: Bitround in python or julia
is `xr_bitround` the bitround from xarray as discussed [here](https://github.com/zarr-developers/numcodecs/issues/298)? For performance I'd suggest to use `round` from BitInformation. I've never actually compared the speed, but recalculation of constants is avoided reaching 16GB/s at which point it's memory-bound (and therefore likely an upper bound). Also gives us more flexibility in case we want to change something on the rounding front. 

__Originally posted by @milankl in https://github.com/observingClouds/bitinformation_pipeline/pull/21#discussion_r843996664__

### PR DESCRIPTION
**Pull request by aaronspring**
- implement `bp.jl_bitround`
- test against `bp.xr_bitround`
- align with `numcodecs.bitround`

Closes #25
deals with #27

### CODE REVIEW

File: bitinformation_pipeline/bitinformation_pipeline.py [hunk 1]
```diff
@@ -160,6 +160,13 @@ def get_inflevel(var, inflevel):
     return keepbits
 
 
+def _jl_bitround(X, keepbits):
+    Main.X = X
+    Main.keepbits = keepbits
+    # wrapping BitInformation.round!($X, int) # what does the $ mean?
```
Comment (milankl) [inline bitinformation_pipeline/bitinformation_pipeline.py:166]: The `$` is only for benchmarking with `@btime` from BenchmarkTools.jl and escapes any calculations outside of the actual function call via string interpolation

Comment (aaronspring) [reply bitinformation_pipeline/bitinformation_pipeline.py:166]: thanks

**Comment by milankl** [comment]
Can you link me to the python code for `numcodecs.bitround`? It's not in master yet right? In Ryan's version it seems that a simple escape is used for keepbits=23 (because no rounding should take place).
```python
def encode(self, buf):
        if self.keepbits == 23:
            return buf
```
I can easily add that for BitInformation.jl too. Because at the moment we have
```julia
julia> bitstring.(A,:split)
10-element Vector{String}:
 "0 01111101 00100011110101011001100"
 "1 01111101 11111100000100010010110"
 "0 01111111 00100110111101011010101"
 "0 01111100 00000101000001101001010"
 "1 01111111 10100000111000010100000"
 "0 01111100 01011110111011000110101"
 "0 01111010 01100000001000101011100"
 "1 01111110 00101101011101001110111"
 "1 01111101 10000010000111001000111"
 "1 01111101 11000011000111110011100"

julia> bitstring.(round(A,22),:split)    # keepbits=22, all correct
10-element Vector{String}:
 "0 01111101 00100011110101011001100"    # no rounding
 "1 01111101 11111100000100010010110"    # no rounding
 "0 01111111 00100110111101011010100"    # round to zero=even (tie)
 "0 01111100 00000101000001101001010"    # no rounding
 "1 01111111 10100000111000010100000"    # no rounding
 "0 01111100 01011110111011000110100"    # round to zero=even (tie)
 "0 01111010 01100000001000101011100"    # no rounding
 "1 01111110 00101101011101001111000"    # round away from zero (with carry)
 "1 01111101 10000010000111001001000"    # round away from zero (with carry)
 "1 01111101 11000011000111110011100"    # no rounding

julia> bitstring.(round(A,23),:split)    # keepbits=23
10-element Vector{String}:
 "0 01111101 00100011110101011001100"    # no rounding, correct
 "1 01111101 11111100000100010010110"    # no rounding, correct
 "0 01111111 00100110111101011010110"    # round away from zero, incorrect
 "0 01111100 00000101000001101001010"    # no rounding, correct
 "1 01111111 10100000111000010100000"    # no rounding, correct
 "0 01111100 01011110111011000110110"    # round away from zero, incorrect
 "0 01111010 01100000001000101011100"    # no rounding, correct
 "1 01111110 00101101011101001111000"    # round away from zero, incorrect
 "1 01111101 10000010000111001001000"    # round away from zero, incorrect
 "1 01111101 11000011000111110011100"    # no rounding, correct
```
Meaning for the edge case of keepbits=23 there's still some rounding away from zero possible, which obviously shouldn't happen. I'll see in a patch release how to best deal with that. The ball here is in my court. But that shouldn't stop us if you want to make BitInformation.round the default.

**Comment by observingClouds** [comment]
This is awesome! I guess we handle keepbit==23 differently in the possible numcodecs implementation and BitInformation. See [here](https://github.com/zarr-developers/numcodecs/pull/299#discussion_r771830593). The suggested numcodecs implementation returns the input as is, while BitInformation should do the same according to [this test](https://github.com/milankl/BitInformation.jl/blob/fa2dd50171a48565ef1dc5a3c7268c19ffe844f1/test/round_nearest.jl#L56). Seems like the issue needs to be fixed upstream.

**Comment by observingClouds** [comment]
Thanks @milankl. The codec is indeed not yet merged and I just have that in my own numcodecs development branch to make it available. With https://github.com/zarr-developers/numcodecs/issues/290 being merged, we could also think of having bitround as an external filter which gets automatically registered with numcodecs. I played a little bit with that [here](https://github.com/observingClouds/bitround).

**Review by observingClouds — APPROVED** [review]

**Pull request merged**